fair_mango/metrics/superset.py

```diff
@@ -0,0 +1,318 @@
+    list
+        A list of dictionaries, each containing the sensitive attributes
+        considered and their corresponding performance evaluation metric
+        results.
+
+    Raises
+    ------
+    AttributeError
+        if data is a pandas dataframe and 'sensitive' parameter is not provided.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...         'gender': ['male', 'male', 'male', 'male', 'female', 'female'],
+    ...         'race': ['white', 'white', 'black', 'black', 'white', 'white'],
+    ...         'real': [1,0,1,0,0,1],
+    ...         'pred': [0,0,1,0,0,1]
+    ... })
+    >>> result = super_set_performance_metrics(
+    ...     data=df,
+    ...     sensitive=['gender', 'race'],
+    ...     real_target=['real'],
+    ...     predicted_target=['pred'],
+    ... )
+    >>> result
+    [
+        {'sensitive': ('gender',),
+        'result': (['real'],
+        [
+            {
+                'sensitive': array(['male'], dtype=object),
+                'selection_rate_in_data': array(0.5),
+                'selection_rate_in_predictions': array(0.25),
+                'accuracy': [0.75],
+                'balanced accuracy': [0.75],
+                'precision': [1.0],
+                'recall': [0.5],
+                'f1-score': [0.6666666666666666],
+                'false_negative_rate': [0.5],
+                'false_positive_rate': [0.0],
+                'true_negative_rate': [1.0],
+                'true_positive_rate': [0.5]
+            },
+            {
+                'sensitive': array(['female'], dtype=object),
+                ...
+                'f1-score': [0.0],
+                'false_negative_rate': [1.0],
+                'false_positive_rate': [0.0],
+                'true_negative_rate': [1.0],
+                'true_positive_rate': [0.0]
+            }
+        ]
+        )
+        }
+    ]
+    """
+    results = []
+    metrics = [SelectionRate, PerformanceMetric, ConfusionMatrix]
+
+    if isinstance(data, Dataset):
+        sensitive = data.sensitive
+        real_target = data.real_target
+        predicted_target = data.predicted_target
+        positive_target = data.positive_target
+        data = data.df
+        if predicted_target == []:
+            predicted_target = None
+
+    if sensitive is None:
+        raise AttributeError(
+            "'sensitive' attribute is required when data is " "pandas dataframe"
+        )
+
+    pairs = chain.from_iterable(
+        combinations(sensitive, r) for r in range(1, len(sensitive) + 1)
+    )
+
+    for pair in pairs:
+        concatenated_results = SelectionRate(
+            data=data,
+            use_y_true=True,
+            sensitive=list(pair),
+            real_target=real_target,
+            predicted_target=predicted_target,
+            positive_target=positive_target,
+            label="selection_rate_in_data",
+        )()
+        for metric in metrics:
+            if metric.__name__ == "SelectionRate":
```

Comment:
avoid using magic methods, there is something called `issubclass` for this job.

Reply:
Using `issubclass` makes `mypy` complain so I found another simple way by doing this check
```python
if metric is SelectionRate:
    ...
```

Reply:
what was the complaint from mypy please?

Reply:
It is this
```
fair_mango/metrics/superset.py:347: error: Never not callable  [misc]
```

Reply:
let's investigate

Reply:
I have no error on my side

Reply:
yep, and up to date. But I had others mypy errors on other files. Run the `mypy` command

Reply:
For me `mypy` passes with no errors in any file.

Reply:
I always try to check it before any commit to avoid including those errors in the commit.

Reply:
<img width="1008" alt="image" src="https://github.com/user-attachments/assets/e4bb548d-1c2a-48f4-ba6e-11f9ffe9c5f3">

what is your version of mypy ? I have 1.9.0

Reply:
I have 1.8.0

Reply:
oh, the hook is indeed 1.8.0, mypy.

Can you research the difference between `metric is FairnessMetric` and `issublass(metric, FairnessMetric)` please ?

Reply:
- `metric is FairnessMetric`: checks if `metric` is exactly `FairnessMetric` and not any other subclasses that inherits `FairnessMetric`. Example: if metric is this class `class DisparateImpact(FairnessMetric)` then the check `metric is FairnessMetric` will return `False`.
- `issublass(metric, FairnessMetric)`: checks `metric` is `FairnessMetric` or any other subclass that inherits `FairnessMetric`. Example:  if metric is this class `class DisparateImpact(FairnessMetric)` then the check `metric is FairnessMetric` will return `True`.

Reply:
In our case using `is` is more suitable. In case in the future for any reason, `SelectionRate` is inherited by another class, we don't want that class to pass in this check because we are specifically interested in `SelectionRate`

Reply:
thanks !

But I think that we should follow the Liskov’s Substitution Principle.

Reply:
Hmm I never heard of this principle. I will look more into it.

Reply:
I prefer the current implementation, using `metric is FairnessMetric`. It seems more clear and direct. For the case we are using it with `SelectionRate`, I doubt that there will be any subclass for `SelectionRate`.